Add more examples to app config (#13204)
elixir-lang/elixir: lib/elixir/pages/anti-patterns/design-anti-patterns.md

["Lucas", "Francisco-da-Matta-Vegi"]
```

> Of course, not all uses of the application environment by libraries are disallowed. For example, the application environment can be coupled with other extensibility mechanisms, such as [behaviours](https://hexdocs.pm/elixir/typespecs.html#behaviours), to remove some of the downsides. For example, if a library needs to parse JSON or CSV files, or perform HTTP requests, the library author may pick one package to use as default but allow its users to swap to different implementations via the application environment.

Author: ```suggestion
Of course, not all uses of the application environment by libraries are incorrect. One example is using configuration to replace a component or a dependency of a library by another that must behave the exact same. Consider a library that needs to parse CSV files. The library author may pick one package to use as default  parser but allow its users to swap to different implementations via the application environment. At the end of the day, choosing a different CSV parser should not change the outcome and library authors can even enforce this by [defining behaviours](../references/typespecs.md#behaviours) with the exact semantics they expect.
```

A possible solution in such cases is to, instead of starting the supervision tree inside the library, the library should provide its own child specifications, so the user of the library starts all necessary processes under its own supervision tree.

You can see this pattern in practice in projects like [Nx](https://github.com/elixir-nx/nx) and [DNS Cluster](https://github.com/phoenixframework/dns_cluster). In order to use these projects (or certain features within them), you must list the necessary processes under your own supervision tree:

```elixir
children = [
  {DNSCluster, query: "my.subdomain"}
]
```

In such cases, if the users of `DNSCluster` need to configure DNSCluster per environment, they can be the ones reading from the application environment, without the library forcing them to:

```elixir
children = [
  {DNSCluster, query: Application.get_env(:my_app, :dns_cluster_query) || :ignore}
]

> ```

Author: ````suggestion
```

Some libraries, such as [Ecto](https://github.com/elixir-ecto/ecto), allow you to pass your application name as an option (called `:otp_app` or similar) and then automatically read the environment from *your* application. While this addresses the issue with the application environment being global, as they read from each individual application, it comes at the cost of some indirection, compared to the example above where users explicitly read their application environment from their own code, whenever desired.
````

["Lucas", "Francisco-da-Matta-Vegi"]
```

Of course, not all uses of the application environment by libraries are incorrect. One example is using configuration to replace a component (or dependency) of a library by another that must behave the exact same. Consider a library that needs to parse CSV files. The library author may pick one package to use as default parser but allow its users to swap to different implementations via the application environment. At the end of the day, choosing a different CSV parser should not change the outcome, and library authors can even enforce this by [defining behaviours](../references/typespecs.md#behaviours) with the exact semantics they expect.

#### Additional remarks: Supervision trees

In practice, libraries may require additional configuration beyond keyword lists. For example, if a library needs to start a supervision tree, how can the user of said library customize its supervision tree? Given the supervision tree itself is global (as it belongs to the library), library authors may be tempted to use the application configuration once more.

> A possible solution in such cases is to, instead of starting the supervision tree inside the library, the library should provide its own child specifications, so the user of the library starts all necessary processes under its own supervision tree.

Reviewer: ```suggestion
One solution is for the library to provide its own child specification, instead of starting the supervision tree itself. This allows the user to start all necessary processes under its own supervision tree, potentially passing custom configuration options during initialization.
```

> You can see this pattern in practice in projects like [Nx](https://github.com/elixir-nx/nx) and [DNS Cluster](https://github.com/phoenixframework/dns_cluster). In order to use these projects (or certain features within them), you must list the necessary processes under your own supervision tree:

Reviewer: ```suggestion
You can see this pattern in practice in projects like [Nx](https://github.com/elixir-nx/nx) and [DNS Cluster](https://github.com/phoenixframework/dns_cluster). These libraries require that you list processes under your own supervision tree:
```
